lib/datasource/datasource.ts

```diff
@@ -0,0 +1,51 @@
+import { ExternalHostError } from '../types/errors/external-host-error';
+import { Http } from '../util/http';
+import type { HttpError } from '../util/http/types';
+import type {
+  DatasourceApi,
+  DigestConfig,
+  GetReleasesConfig,
+  ReleaseResult,
+} from './types';
+
+export abstract class Datasource implements DatasourceApi {
+  protected constructor(public readonly id: string) {
+    this.http = new Http(id);
+  }
+
+  caching: boolean;
+
+  defaultConfig: Record<string, unknown>;
+
+  customRegistrySupport: boolean;
+
+  defaultRegistryUrls: string[];
+
+  defaultVersioning: string;
+
+  registryStrategy: 'first' | 'hunt' | 'merge';
+
+  protected http: Http;
+
+  abstract getReleases(
+    getReleasesConfig: GetReleasesConfig
+  ): Promise<ReleaseResult | null>;
+
+  getDigest?(config: DigestConfig, newValue?: string): Promise<string>;
+
+  // eslint-disable-next-line class-methods-use-this
+  handleSpecificErrors(err: HttpError): void {}
```

Comment:
We should define an order to we chave consistency. 🤔

Reply:
An order for what?

Reply:
member type order and alphabetic naming. maybe we need to open an issue to discuss best practices for us?

Reply:
Not sure it's that big a deal right now. Maybe later if this becomes more busy.

For now I've sorted alphabetically, with `protected http` at the bottom as it's a different modifier.

Reply:
you are right, I've opened an issue to track this, so we can make a desision in future. ref #9948